xds/internal/balancer/edsbalancer/balancergroup.go

```diff
@@ -56,29 +109,74 @@ type pickerState struct {
 //     - sub-pickers are grouped into a group-picker
 //     - aggregated connectivity state is the overall state of all pickers.
 //  - resolveNow
+//
+// Sub-balancers are only built when the balancer group is started. If the
+// balancer group is closed, the sub-balancers are also closed. And it's
+// guaranteed that no updates will be sent to parent ClientConn from a closed
+// balancer group.
 type balancerGroup struct {
-	cc balancer.ClientConn
-
-	mu           sync.Mutex
-	idToBalancer map[internal.Locality]balancer.Balancer
-	scToID       map[balancer.SubConn]internal.Locality
-	loadStore    lrs.Store
+	cc        balancer.ClientConn
+	loadStore lrs.Store
 
-	pickerMu sync.Mutex
-	// All balancer IDs exist as keys in this map. If an ID is not in map, it's
-	// either removed or never added.
+	outgoingMu         sync.Mutex
+	outgoingStarted    bool
+	idToBalancerConfig map[internal.Locality]*balancerConfig
+
+	// incomingMu and pickerMu are to make sure this balancer group doesn't send
+	// updates to cc after it's closed.
+	//
+	// We don't share the mutex to avoid deadlocks (e.g. a call to sub-balancer
```

Comment:
Having multiple mutexes also seem like a recipe for a deadlock. Is there a lock acquisition order that is defined?

Reply:
And I also don't like the multiple mutex much. Another solution is to have one goroutine to synchronize all the updates (like [this](https://github.com/grpc/grpc-go/blob/74f33a446dd3c95a64a1e8fe7916f73ec70b0a6c/balancer_conn_wrappers.go#L113)).
This have the overhead of one more goroutine though.